weather: add blurred color gradient, changes based on weather

matildepark opened the pull request:
As per mocks.

Colours were indeterminate; I did my best intuitive stab at it with Indigo colours, but the mocks didn't use Indigo colours either. Thus the @urcades review.

<img width="653" alt="Screen Shot 2020-02-18 at 8 55 31 PM" src="https://user-images.githubusercontent.com/20846414/74795806-e5aa4480-5294-11ea-9f3e-219ee8492698.png">

urcades commented:
@matildepark :

<img width="1502" alt="Screen Shot 2020-02-18 at 11 03 57 PM" src="https://user-images.githubusercontent.com/1195363/74800934-0e393b00-52a3-11ea-8ac8-58fcc14fa880.png">

Damn, this is nice — will look through the individual color expressions now. Do you have any idea what that space is in the upper left?

matildepark commented:
> Do you have any idea what that space is in the upper left?

That's the Links tile, which is broken on `os1-rc` pull because the compiled version in the repo is before some tile-specific changes.

urcades commented:
> > Do you have any idea what that space is in the upper left?
> 
> That's the Links tile, which is broken on `os1-rc` pull because the compiled version in the repo is before some tile-specific changes.

Gotcha, good to know — will leave that out of review

matildepark commented:
If the gradient is too apparent, feel free to increase the `blur` amount — also let me know if you have a recommendation for the gradient settings ... you should see both of those fairly easily in the inspector, they're inline.

urcades commented:
<img width="1502" alt="Screen Shot 2020-02-18 at 11 15 35 PM" src="https://user-images.githubusercontent.com/1195363/74801448-99ff9700-52a4-11ea-9c5d-25ffe4956413.png">

fwiw, gradient rendering on chrome is so, so, so much better. I've been playing around with a few blur options and something around the neighborhood of 80px is looking really good for the overcast tile. 

Currently trying to figure out how to override the weather display and get a sense of what everything else looks like.

matildepark commented:
Open React DevTools, mouse over Weather component, crack open props -> data -> daily -> icon. Edit it to one of the options in the switch statement and press `Enter`.

![image](https://user-images.githubusercontent.com/20846414/74801896-22cb0280-52a6-11ea-9d15-fc872d345687.png)

urcades commented:
> > Currently trying to figure out how to override the weather display and get a sense of what everything else looks like.
> 
> Open React DevTools, mouse over Weather component, crack open props -> data -> daily -> icon. Edit it to one of the options in the switch statement and press `Enter`.
> 
> ![image](https://user-images.githubusercontent.com/20846414/74801896-22cb0280-52a6-11ea-9d15-fc872d345687.png)

I'm an idiot: on new machine and didn't have any of that react tooling/toolchain installed ... forgot this stuff isn't built into chrome. Reviewing weather types now.

urcades commented:
Actually, now that I've posted the above, it looks like your browser renders it fine — I'm using chrome to view this PR, and I can't seem to shake off that blur.

matildepark commented:
Chrome / Brave:

<img width="380" alt="Screen Shot 2020-02-19 at 1 14 04 PM" src="https://user-images.githubusercontent.com/20846414/74862235-c9042000-5319-11ea-99c6-8e4828439fb7.png">

urcades commented:
Holy shit lol, I literally just discovered what the issue is — at varying browser zoom levels, it appears as though that undesired gradient makes an appearance that can't really be controlled:

![Kapture 2020-02-19 at 13 11 54](https://user-images.githubusercontent.com/1195363/74862193-b2f65f80-5319-11ea-8dde-7a4c35922cad.gif)

urcades commented:
I think we can ignore my request for changes

matildepark commented:
Ah, ok, I can replicate that zoom stuff. That seems out of scope ... mind stamping?

urcades commented:
@matildepark oh, 80px would be good — in the future do you want me to directly edit PR source with small changes like that? lmk whatever makes your work easier.

matildepark commented:
Alrighty, ready for @loganallenc stamp. 

> in the future do you want me to directly edit PR source with small changes like that?

It doesn't take much time to amend, don't worry about it.

loganallenc commented on pkg/interface/weather/tile/tile.js:
```diff
@@ -62,18 +46,22 @@ export default class WeatherTile extends Component {
         style={{
           width: 126,
           height: 126,
+          background: `linear-gradient(135deg, ${weatherStyle.gradient1} 0%,` +
```
fascinating. I like it!

loganallenc commented on pkg/interface/weather/tile/tile.js:
```diff
@@ -62,18 +46,22 @@ export default class WeatherTile extends Component {
 
-  renderWrapper(child) {
+  renderWrapper(child,
+    weatherStyle = { gradient1: "white", gradient2: "white", text: "black" }
```
weird that you aren't passing these in as separate arguments, as an object is *really* unstructured

loganallenc commented on pkg/interface/weather/tile/tile.js:
```diff
@@ -182,7 +178,91 @@ export default class WeatherTile extends Component {
+        text: ""
+      };
+
+      switch(data.daily.icon) {
```
Put this whole switch statement in its own function. General rule of thumb: one function shouldn't be more than 40 lines

matildepark commented:
I had some leftover messy code I cleaned up; refactored the style change into its own function; added some comments to help contributors navigate the code. 

I could replicate @urcades' issue myself by just resizing the window around; seemingly [there's a Chrome bug about this](https://stackoverflow.com/questions/57736567/rendering-flickering-glitch-issue-in-chrome-using-css-backdrop-filterblur). 

I still fixed the issue with the black gradient overlay by passing the bordering behaviour to child components (removing it from the paint, since it was taking up 2px of the 126px, thus making the child 124px; placing it in the child stops resizing the backdrop behind the child and thus removing the black gradient overlay).
